File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation
  ( proteins,
  )
where

import Data.List.Split (chunksOf)
import Data.Map.Strict (Map, (!?))
import qualified Data.Map.Strict as Map

proteins :: String -> Maybe [String]
proteins = fmap (takeWhile (/= "STOP")) . mapM (translations !?) . chunksOf 3

translations :: Map String String
translations =
  Map.fromList
    . concatMap (\(codons, protein) -> [(codon, protein) | codon <- codons])
    $ [ (["AUG"], "Methionine"),
        (["UGG"], "Tryptophan"),
        (["UUU", "UUC"], "Phenylalanine"),
        (["UUA", "UUG"], "Leucine"),
        (["UCU", "UCC", "UCA", "UCG"], "Serine"),
        (["UAU", "UAC"], "Tyrosine"),
        (["UGU", "UGC"], "Cysteine"),
        (["UAA", "UAG", "UGA"], "STOP")
      ]

File: go/protein-translation/protein_translation.go
package protein

import (
	"errors"
	"sort"
)

var ErrStop = errors.New("STOP")

var ErrInvalidBase = errors.New("invalid base")

func FromRNA(rna string) ([]string, error) {
	codons := make([]string, 0)
	for {
		if len(rna) < 3 {
			break
		}

		aminoAcid, err := FromCodon(rna[:3])
		if err == ErrInvalidBase {
			return []string{}, err
		}
		if err == ErrStop {
			rna = ""
			break
		}

		codons = append(codons, aminoAcid)
		rna = rna[3:]
	}

	if rna != "" {
		return []string{}, ErrInvalidBase
	}

	return codons, nil
}

func FromCodon(codon string) (string, error) {
	protein, ok := aminoAcids()[codon]
	if !ok {
		return "", ErrInvalidBase
	}

	if protein == "STOP" {
		return "", ErrStop
	}

	return protein, nil
}

func aminoAcids() map[string]string {
	codons := make(map[string]string)
	codons["AUG"] = "Methionine"
	codons["UUU"] = "Phenylalanine"
	codons["UUC"] = codons["UUU"]
	codons["UUA"] = "Leucine"
	codons["UUG"] = codons["UUA"]
	codons["UCU"] = "Serine"
	codons["UCC"] = codons["UCU"]
	codons["UCA"] = codons["UCU"]
	codons["UCG"] = codons["UCU"]
	codons["UAU"] = "Tyrosine"
	codons["UAC"] = codons["UAU"]
	codons["UGU"] = "Cysteine"
	codons["UGC"] = codons["UGU"]
	codons["UGG"] = "Tryptophan"
	codons["UAA"] = "STOP"
	codons["UAG"] = codons["UAA"]
	codons["UGA"] = codons["UAA"]

	return codons
}

func reverse[T comparable](s []T) {
	sort.SliceStable(s, func(i, j int) bool {
		return i > j
	})
}
